### PR TITLE
Initial trait definition for `relocatable`

### PR DESCRIPTION
**Pull request by isidorostsa**
This is a draft pr to implement the basic traits for P1144 (relocatable).

I defined a meta function (`is_relocatable`) and `pointer_relocate_category` allongside the existing `pointer_copy_category` and `pointer_move_category`. 

A type is `relocatable` if it is not tied to a specific location in memory.  (An example object that is not relocatable is std::mutex)

The aim of this branch is to implement the trivially relocatable trait and the faster data transfer algorithms it allows us to use.

To move forward I ask if this is the right file to the `is_relocatable` definitions.

I believe the `trivially_copyable_pointer` part of this file focuses on if an iterator is suitable for a contiguous `memcpy`, and not in determining if the underlying object is `trivially_copyable`; it uses `std::is_trivially_copyable` for that. At the moment however there is no `std::is_trivially_relocatable` or `std::is_relocatable`, so these have to be implemented first.

At the moment I have placed `hpx::is_relocatable` in this file, should I place it somewhere else?  

Thanks!

### CODE REVIEW

File: libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp
Comment (isidorostsa) [inline libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp:143]: This is the part I should probably move to another file

File: libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp:154]: Should we make this to not use double negation? That would mean something like:
```suggestion
            bool IsContiguous = iterators_are_contiguous_v<Source, Dest>>
```

File: libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp:197]:
```suggestion
    template <typename Source, typename Dest, typename Enable = void>
    struct pointer_relocate_category : detail::pointer_relocate_category<Source, Dest>
    {
    };
```

File: libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp
Comment (isidorostsa) [reply libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp:154]: agreed, this happens in multiple places in the file, should I change the pre-exising stuff too?

Comment (hkaiser) [reply libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp:154]: > agreed, this happens in multiple places in the file, should I change the pre-exising stuff too?

Sure, if it's not too much of a bother.

Comment (isidorostsa) [reply libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp:154]: Not at all, what do you think about the file location of `hpx::is_relocatable`?

Comment (hkaiser) [reply libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp:154]: This could be a new file in this directory, I think.

Comment (isidorostsa) [reply libs/core/algorithms/include/hpx/algorithms/traits/pointer_category.hpp:154]: Thanks, I will work on that and get back to you

**Comment by gonidelis** [comment]
Maybe adding tests would help :)

File: libs/core/algorithms/include/hpx/algorithms/traits/is_relocatable.hpp
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/algorithms/traits/is_relocatable.hpp:9]:
```suggestion
#include <type_traits>
```
Both, `is_destructible` and `is_move_constructible` are defined in `<type_traits>`

File: libs/core/algorithms/include/hpx/algorithms/traits/is_relocatable.hpp
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/algorithms/traits/is_relocatable.hpp:1]: Please use your copyright

Comment (isidorostsa) [reply libs/core/algorithms/include/hpx/algorithms/traits/is_relocatable.hpp:1]: Fixed!

File: libs/core/algorithms/include/hpx/algorithms/traits/is_relocatable.hpp
Comment (isidorostsa) [reply libs/core/algorithms/include/hpx/algorithms/traits/is_relocatable.hpp]: Changed it

**Review by hkaiser — APPROVED** [review]
LGTM, thanks! However, please fix the clang-format and cmake-format issues.

File: libs/core/algorithms/CMakeLists.txt
Comment (hkaiser) [inline libs/core/algorithms/CMakeLists.txt:229]: Yu don't need this entry. This is only for compatibility with old locations of headers that have been moved around.

**Comment by gonidelis** [comment]
I like it. `is_relocatable` = `is_move_constructible` && `is_destructible` is a simple idea on which we can build upon.

**Comment by isidorostsa** [comment]
> Maybe adding tests would help :)

Thanks, I will add some assertion tests.

**Comment by hkaiser** [comment]
@isidorostsa Wrt testing: I think having a couple of tests would be nice, most likely compile tests using `static_assert` that verify the traits really do what you expect. Please have a look at the exiting testing infrastructure for clues on how to do that.

**Comment by isidorostsa** [comment]
> @isidorostsa Wrt testing: I think having a couple of tests would be nice, most likely compile tests using `static_assert` that verify the traits really do what you expect. Please have a look at the exiting testing infrastructure for clues on how to do that.

Thanks for the hints, I will add the tests later today

**Comment by hkaiser** [comment]
@isidorostsa Please don't think we're ignoring this. Let's wait for the release of HPX V1.9.0 (next week) before merging.

**Comment by isidorostsa** [comment]
> @isidorostsa Please don't think we're ignoring this. Let's wait for the release of HPX V1.9.0 (next week) before merging.

That's great to hear! 

I wanted to ask you, was I correct to add my name to the copyrights list section of all the files I changed?  
Also is there a specific HPX macro for static assertions or is `static_assert` okay?

**Comment by hkaiser** [comment]
> wanted to ask you, was I correct to add my name to the copyrights list section of all the files I changed?

The rule of thumb we usually apply is that one should add his/her name to the copyrights whenever the person touched on ~20% of the file.

> Also is there a specific HPX macro for static assertions or is `static_assert` okay?

`static_assert()` is just fine.

**Comment by hkaiser** [comment]
@isidorostsa could you please rebase this onto master, now that the release is out?

**Comment by isidorostsa** [comment]
> @isidorostsa could you please rebase this onto master, now that the release is out?

I think I did it, but could you confirm that as this is not something I've done before? Thanks!

**Comment by hkaiser** [comment]
> > @isidorostsa could you please rebase this onto master, now that the release is out?
> 
> I think I did it, but could you confirm that as this is not something I've done before? Thanks!

LGTM, thanks.

**Review by hkaiser — APPROVED** [review]
I think this is good to be merged. LGTM, thanks!

**Comment by hkaiser** [comment]
bors merge

**Comment by hkaiser** [comment]
@isidorostsa Congratulations, your first PR to the HPX repository has just been merged! As a 'thank you' we offer a free STE||AR-Group t-shirt to all of our first-time contributors. If you are interested in receiving one, please get back to me directly so we can set up the delivery.

**Comment by hkaiser** [comment]
> @isidorostsa after merging your PR, compilation is failing across the board, see for instance here: https://github.com/STEllAR-GROUP/hpx/actions/runs/4893500203/jobs/8736534493. Would you have the time to investigate?

Please see https://github.com/STEllAR-GROUP/hpx/pull/6233 for a possible fix.

**Pull request merged**
